File: 动态规划/1039/1039.cpp
class Solution {
public:
	int minScoreTriangulation(vector<int>& A) {
		int dp[55][55], len = A.size();
		memset(dp, 0, sizeof(dp));
		for (int i = 0; i < len - 2; i++) {
			dp[i][i + 2] = A[i] * A[i + 1] * A[i + 2];
		}
		for (int k = 3; k < len; k++) {
			for (int i = 0; i < len - k; i++) {
				int j = i + k, temp = 1e9;
				for (int l = i + 1; l < j; l++) {
					temp = min(temp, dp[i][l] + dp[l][j] + A[i] * A[j] * A[l]);
				}
				dp[i][j] = temp;
			}
		}
		return dp[0][len - 1];
	}
};

File: 动态规划/1039/1039.go
func minScoreTriangulation(values []int) int {
	n := len(values)
	dp := make([][]int, n)
	for i := 0; i < n; i++ {
		dp[i] = make([]int, n, n)
	}

	for i := 0; i < n-2; i++ {
		dp[i][i+2] = values[i] * values[i+1] * values[i+2]
	}

	//重要的是构造连续的区间表示，以此来应用区间dp
	for k := 3; k < n; k++ {
		for start := 0; start < n-k; start++ {
			end := start + k
			dp[start][end] = math.MaxInt
			for p := start + 1; p < end; p++ {
				temp := dp[start][p] + dp[p][end] + values[start]*values[p]*values[end]
				if temp < dp[start][end] {
					dp[start][end] = temp
				}
			}
		}
	}

	return dp[0][n-1]
}